// react-image-magnifiers/dist/GlassRenderer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _utils = _interopRequireDefault(require("./utils"));

var _styles = _interopRequireDefault(require("./styles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var GlassRenderer = function GlassRenderer(props) {
  var itemRef = props.itemRef,
      itemPosition = props.itemPosition,
      activePosition = props.activePosition,
      elementDimensions = props.elementDimensions,
      itemDimensions = props.itemDimensions,
      active = props.active,
      imageSrc = props.imageSrc,
      largeImageSrc = props.largeImageSrc,
      imageAlt = props.imageAlt,
      magnifierBorderSize = props.magnifierBorderSize,
      magnifierBorderColor = props.magnifierBorderColor,
      magnifierBackgroundColor = props.magnifierBackgroundColor,
      square = props.square,
      magnifierSize = props.magnifierSize,
      magnifierOffsetX = props.magnifierOffsetX,
      magnifierOffsetY = props.magnifierOffsetY,
      renderOverlay = props.renderOverlay;
  var isActive = itemDimensions.width > elementDimensions.width && active;

  var magnifierSizeNum = _utils.default.convertWidthToPx(magnifierSize, elementDimensions.width);

  var positionOffset = magnifierSizeNum / 2;
  var position = {
    x: itemPosition.x - activePosition.x + positionOffset - magnifierBorderSize,
    y: itemPosition.y - activePosition.y + positionOffset - magnifierBorderSize
  };
  var divPosition = {
    x: activePosition.x - positionOffset + magnifierOffsetX,
    y: activePosition.y - positionOffset + magnifierOffsetY
  };
  var borderRadius = square ? "0" : "50%";
  return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("img", {
    src: imageSrc,
    alt: imageAlt,
    style: {
      width: "100%",
      display: "block",
      boxSizing: "border-box"
    }
  }), _react.default.createElement("div", {
    style: _objectSpread({}, _styles.default.getZoomContainerStyle(magnifierSizeNum, magnifierSizeNum, true), {
      visibility: !active ? "hidden" : "visible",
      borderRadius: borderRadius,
      zIndex: "1",
      border: "".concat(magnifierBorderSize, "px solid ").concat(magnifierBorderColor),
      transform: "translate(".concat(divPosition.x, "px, ").concat(divPosition.y, "px)"),
      backgroundColor: magnifierBackgroundColor,
      backgroundClip: "padding-box"
    })
  }, _react.default.createElement("img", {
    ref: itemRef,
    src: largeImageSrc || imageSrc,
    alt: imageAlt,
    style: _styles.default.getLargeImageStyle(position.x, position.y, isActive)
  })), renderOverlay ? renderOverlay(active) : null);
};

var _default = GlassRenderer;
exports.default = _default;
