// react-image-magnifiers/dist/MagnifierRenderer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _styles = _interopRequireDefault(require("./styles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MagnifierRenderer = function MagnifierRenderer(props) {
  var itemPosition = props.itemPosition,
      active = props.active,
      elementDimensions = props.elementDimensions,
      itemDimensions = props.itemDimensions,
      imageSrc = props.imageSrc,
      largeImageSrc = props.largeImageSrc,
      imageAlt = props.imageAlt,
      itemRef = props.itemRef,
      renderOverlay = props.renderOverlay;
  var isActive = itemDimensions.width > elementDimensions.width && active;
  return _react.default.createElement("div", {
    style: {
      position: "relative"
    }
  }, _react.default.createElement("img", {
    style: {
      width: "100%",
      display: "block"
    },
    src: imageSrc,
    alt: imageAlt
  }), _react.default.createElement("div", {
    style: _styles.default.getZoomContainerStyle(elementDimensions.width, elementDimensions.height, true)
  }, _react.default.createElement("img", {
    style: _styles.default.getLargeImageStyle(itemPosition.x, itemPosition.y, isActive),
    src: largeImageSrc || imageSrc,
    alt: imageAlt,
    ref: itemRef
  })), renderOverlay ? renderOverlay(active) : null);
};

var _default = MagnifierRenderer;
exports.default = _default;
